src/source/geojson_source_diff.ts

```
@@ -0,0 +1,139 @@
+    if (data.type === 'Feature') {
+        return getFeatureId(data, promoteId) != null;
+    }
+
+    // a feature collection can be updated if every feature has an id, and the ids are all unique
+    // this prevents us from silently dropping features if ids get reused
+    if (data.type === 'FeatureCollection') {
+        const seenIds = new Set<GeoJSONFeatureId>();
+        for (const feature of data.features) {
+            const id = getFeatureId(feature, promoteId);
+            if (id == null) {
+                return false;
+            }
+
+            if (seenIds.has(id)) {
+                return false;
+            }
+
+            seenIds.add(id);
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+export function toUpdateable(data: UpdateableGeoJSON, promoteId?: string) {
+    const result = new Map<GeoJSONFeatureId, GeoJSON.Feature>();
+    if (data == null) {
+        // empty result
+    } else if (data.type === 'Feature') {
+        result.set(getFeatureId(data, promoteId)!, data);
+    } else {
+        for (const feature of data.features) {
+            result.set(getFeatureId(feature, promoteId)!, feature);
+        }
+    }
+
+    return result;
+}
+
+// mutates updateable
+export function applySourceDiff(updateable: Map<GeoJSONFeatureId, GeoJSON.Feature>, diff: GeoJSONSourceDiff, promoteId?: string): void {
+    if (diff.removeAll) {
+        updateable.clear();
+    }
+
+    if (diff.removed) {
+        for (const id of diff.removed) {
+            updateable.delete(id);
+        }
+    }
+
+    if (diff.add) {
+        for (const feature of diff.add) {
+            const id = getFeatureId(feature, promoteId);
+
+            if (id != null) {
+                updateable.set(id, feature);
+            }
+        }
+    }
+
+    if (diff.update) {
+        for (const update of diff.update) {
+            let feature = updateable.get(update.id);
+
+            if (feature == null) {
```

do we want to assert this is non-null or warn if it's null instead of silently discarding the update?

---

@HarelM suggested that it just do as much as possible without throwing errors. It'll definitely be hard to debug updates not working when its used incorrectly though.